profiler/src/main/java/com/splunk/opentelemetry/profiler/context/SpanContextualizer.java

```diff
@@ -0,0 +1,137 @@
+
+import static com.splunk.opentelemetry.profiler.context.StackDescriptorLineParser.CANT_PARSE_THREAD_ID;
+
+import com.splunk.opentelemetry.profiler.events.ContextAttached;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.jfr.consumer.RecordedThread;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Keeps track of span scope changes and, when applicable, can wrap the RecordedEvent with span
+ * context information. This class is not thread safe.
+ */
+public class SpanContextualizer {
+
+  private static final Logger logger = LoggerFactory.getLogger(SpanContextualizer.class);
+
+  private final Pattern lineSplitter = Pattern.compile("\n");
+  private final ThreadContextTracker threadContextTracker = new ThreadContextTracker();
+  private final StackDescriptorLineParser descriptorParser = new StackDescriptorLineParser();
+
+  /**
+   * This updates the tracked thread context for a given span. This must only be called with
+   * ContextAttached events.
+   */
+  public void updateContext(RecordedEvent event) {
+    if (event.getByte("direction") == ContextAttached.IN) {
+      threadContextStarting(event);
+    } else {
+      threadContextEnding(event);
+    }
+  }
+
+  /** Parses thread info from the raw stack string and links it to a span (if available). */
+  public StackToSpanLinkage link(String rawStack) {
+    List<String> frames = Arrays.asList(lineSplitter.split(rawStack));
+    if (frames.size() < 2) {
+      // Many GC and other VM threads don't actually have a stack...
+      return StackToSpanLinkage.withoutLinkage(rawStack);
+    }
+    long threadId = descriptorParser.parseThreadId(frames.get(0));
+    if (threadId == CANT_PARSE_THREAD_ID) {
+      return StackToSpanLinkage.withoutLinkage(rawStack);
+    }
+    return link(rawStack, threadId);
+  }
+
+  /** Links the raw stack with the span info for the given thread. */
+  StackToSpanLinkage link(String rawStack, long threadId) {
+    List<SpanLinkage> inFlightSpansForThisThread =
+        threadContextTracker.getInFlightSpansForThread(threadId);
+
+    if (inFlightSpansForThisThread.isEmpty()) {
+      // We don't know about any in-flight spans for this stack
+      return StackToSpanLinkage.withoutLinkage(rawStack);
+    }
+
+    // This thread has a span happening, augment with span details
+    if (inFlightSpansForThisThread.size() > 1) {
+      logger.debug("!! Nested spans detected: We will only use the last span for now...");
+      logger.debug(
+          "traceIds for thread -> {}",
+          inFlightSpansForThisThread.stream()
+              .map(SpanLinkage::getTraceId)
+              .collect(Collectors.joining(" ")));
+      logger.debug(
+          "spans for thread -> {}",
+          inFlightSpansForThisThread.stream()
+              .map(SpanLinkage::getSpanId)
+              .collect(Collectors.joining(" ")));
```

Comment:
Oh, one more thing that I just noticed: WDYT about guarding this with `log.isDebugEnabled()`? Iterating over all those streams shouldn't happen if we don't log debug